New: Puppet provisioning (experimental support). Refs #37

lingxiaoyang opened the pull request:
This pull request adds puppet provisioning support for LXDock, tested with [this lxdock.yml and site.pp](https://gist.github.com/lingxiaoyang/83dc9291073876342bbb6f54488f685f). 

New module: `lxdock.provisioners.puppet.PuppetProvisioner`.

It is implemented with the intention to mimic Vagrant's Puppet provisioning. The options are taken from [Vagrant doc](https://www.vagrantup.com/docs/provisioning/puppet_apply.html#options) and the usage of `puppet apply` command is taken from [Vagrant implementation](https://github.com/mitchellh/vagrant/blob/9c299a2a357fcf87f356bb9d56e18a037a53d138/plugins/provisioners/puppet/provisioner/puppet.rb#L173).

New public methods: `Guest.copy_file(host_path, guest_path)` and `Guest.copy_directory(host_path, guest_path)` (uses `pathlib` that exists since Python 3.4)

lingxiaoyang commented on lxdock/provisioners/ansible.py:
```diff
@@ -40,10 +40,31 @@ def provision(self):
+    def setup_guest_arch(self):
+        # On archlinux guests we have to ensure that sshd is started!
+        # https://bbs.archlinux.org/viewtopic.php?id=205367
+        self.guest.run(['systemctl', 'enable', 'sshd.socket', '--now'])
```
Question: I am repeating this "ensure sshd" code in both `ansible.py` and `puppet.py`. Is there a better way to do this? For example, I was also thinking to put it as `ArchLinuxGuest.ensure_sshd()`.

hsoft replied:
I think it would be a good idea, yes.

lingxiaoyang commented on lxdock/provisioners/puppet.py:
```diff
@@ -0,0 +1,122 @@
+
+        # Copy manifest file
+        self.host.run(['scp'] + ssh_options + [
+            self.options['manifest_file'], 'root@{}:{}'.format(ip, self._guest_manifest_file)])
```
Question: is there a better way to copy files/folders?

hsoft replied:
Yes. See https://github.com/lxdock/lxdock/blob/master/lxdock/network.py#L106

lingxiaoyang replied:
@hsoft Thank you!

lingxiaoyang commented on lxdock/provisioners/puppet.py:
```diff
@@ -0,0 +1,122 @@
+    def provision(self):
+        """ Performs the provisioning operations using puppet. """
+        ip = get_ipv4_ip(self.guest.lxd_container)
+
```
Note to myself: check if `puppet` exists before proceeding. If it doesn't exist, raise an error to let user know.

Commit:
New: Puppet provisioning (experimental support). Refs #37

Other changes:
1. Add lxdock.conf.schema.get_schema() for testing purposes.
2. Improve the autodiscovery of provisioner schemas:
   (1) support composite voluptuous schema (Any, All, ...);
   (2) prevent "fall-through" error message when the provisioner schema is not satisfied:
         Previously, if the target schema (uniquely determined by "type") is not satisfied,
         the config will continue to be compared with remaining schemas and eventually LXDock
         throws an error message that is totally unrelated to target schema (because of Any).
3. Do the bare bone setups for the provisioners before calling any .provision() method, to prevent
   package clashing if the user defines other provisioners that prepare required packages for a
   later provisioner. Let's ensure that user-defined operations always override LXDock's default.
4. Add `apt-get update` in DebianGuest.install_packages before running `apt-get install`.

Thanks hsoft and ganto for reviewing.

hsoft commented:
This looks well written to me, good job! I haven't tested all the options because I don't have a non-trivial puppet config handy here. @lingxiaoyang, was this code tested with non-trivial puppet provisioning? If yes, I think we can merge as-is.

lingxiaoyang commented:
@hsoft  I'll test further.

ganto commented on docs/provisioners/puppet.rst:
```diff
@@ -0,0 +1,103 @@
+  image: ubuntu/xenial
+
+  provisioning:
+    - type: ansible
```
Looks like here is still a typo

lingxiaoyang replied:
@ganto  Please advise?

lingxiaoyang replied:
Fixed in 06b87e8.

ganto commented:
Shouldn't this be `type: puppet`?

lingxiaoyang commented:
@ganto  Haha, you are right. I didn't notice...

Thank you!

lingxiaoyang commented:
Reminder to myself - TODO: non-trivial puppet provisioning test.

lingxiaoyang commented:
Problems found when testing this puppet recipe of LAMP setting: https://github.com/hollsk/puppet-lamp

1. Need to support `manifests_path`;
2. Folder copy is too slow when there are many files, particularly the module folder.

lingxiaoyang commented:
Finally updated in 65dcb23!

What I have changed since the previous review:
1. After testing with a few Vagrant provisioning setups, I finally chose to copy the entire logic of Vagrant counterpart. In module [`lxdock.provisioners.puppet`](https://github.com/lxdock/lxdock/pull/69/commits/65dcb23c4cc08426779af363253c48166a436825#diff-7652a936db8ff9a996b029c58583eff7), you will see more references to Vagrant code.
2. `Guest.copy_directory` is now implemented by creating/extracting a tar file, because recursively calling `Guest.copy_file` is way too slow.
3. The doc is rewritten.

Other changes not directly related to Puppet provisioning itself (I can split them into multiple commits, if needed):
1. Add `lxdock.conf.schema.get_schema()` for testing purposes.
2. Improve the autodiscovery of provisioner schemas:
   1. support composite voluptuous schema (`Any`, `All`, ...) to be defined in provisioner class.
   2. prevent "fall-through" error message when the provisioner schema is not satisfied:
         - Previously, if the target schema (uniquely determined by "type") is not satisfied, the config will continue to be compared with remaining schemas and eventually LXDock throws an error message that is totally unrelated to target schema (because of `Any`).
3. Do the bare bone setups for the provisioners before calling any `.provision()` method, to prevent
   package clashing if the user defines other provisioners that prepare required packages for a
   later provisioner. Let's ensure that user-defined operations always override LXDock's default.

I have tested with:
- https://github.com/hollsk/puppet-lamp: "manifest" mode + Puppet modules
- https://github.com/mikespokefire/puppet-hiera-test: "manifest" mode + modules + hiera config + facter
- https://github.com/eddumelendez/vagrant-puppet-examples/tree/master/jenkins and https://github.com/eddumelendez/vagrant-puppet-examples/tree/master/mongodb: "manifest" mode + modules
- https://gitlab.berlinonline.net/puppets/deploy-template: "environment" mode + binary path + facter + env variables. `puppet apply` was successfully kicked off (the environment was correctly found) but this repo had too many dependencies so the provisioning didn't go through. 

I didn't find more valid Vagrant projects that use "environment" mode. That's why I put "experimental support" in docs as it still needs improvement, but I think it's fairly usable now.

hsoft commented:
Once again, thanks for this heavy work, much appreciated. I've tested the puppet-lamp scenario myself with this lxdock config:

```
name: puppet-lamp
image: ubuntu/trusty
privileged: true

provisioning:
  - type: puppet
    manifests_path: puppet/manifests
    manifest_file: base.pp
    module_path: puppet/modules
```

There are two problems I encountered.

First, the fact that an `apt update` had never ran prevented puppet from being properly installed. I know that it's not strictly related to this PR, but it's the first time it causes us problems. What would you say about adding an `apt update` run in our debian's `install_packages()`?

Second, `puppet apply` fails here with this message:

```
Error: Puppet::Parser::AST::Resource failed with error ArgumentError: Could not find declared class apache at /.lxdock.d/puppet/manifests/lamp.pp:11 on node puppet-lamp-default-1451369.lxd
Wrapped exception:
Could not find declared class apache
Error: Puppet::Parser::AST::Resource failed with error ArgumentError: Could not find declared class apache at /.lxdock.d/puppet/manifests/lamp.pp:11 on node puppet-lamp-default-1451369.lxd
```

Did you get a similar error @lingxiaoyang ?

lingxiaoyang commented:
@hsoft Thank you for testing! I confirm the first problem (`apt update`) after destroying and rebuilding the container. I think it's necessary to `update` first. I'll add it.

I didn't replicate problem 2. I attach my logs here:
[puppet-lamp-test.txt](https://github.com/lxdock/lxdock/files/1078739/puppet-lamp-test.txt)

I used the same lxdock file as in your comment. After `apt update`, I got:

````
root@puppet-lamp-default-13681098:~# puppet --version
3.4.3
````

Do you have the same version in the container?

hsoft commented:
Yes, same version. Since I'm the only one getting the error, I'll do the debugging.

lingxiaoyang commented:
@hsoft  You need `git submodule init` and `git submodule update`.

hsoft commented:
Oh, silly me, of course! retrying now.

hsoft commented:
Yup, it works fine now. For the more complex setups, I'll trust that they work without actually trying them.

Code-wise, this looks good to me. I'd merge. @ellmetha you want to take a look first?

ellmetha commented:
@hsoft it's OK for me. I didn't have time to test the latest changes in this PR but if you think that we're good to go with these changes, then so do I. 😉 

@lingxiaoyang good work btw!
